// @kernel-js/support/es6/Helpers/index.js
/*----------------------------------------------
 *                    Lodash
 *---------------------------------------------- */
var camelCase = require('lodash/camelCase');
var capitalize = require('lodash/capitalize');
var find = require('lodash/find');
var floor = require('lodash/floor');
var get = require('lodash/get');
var has = require('lodash/has');
var hasIn = require('lodash/hasIn');
var head = require('lodash/head');
var keysIn = require('lodash/keysIn');
var last = require('lodash/last');
var set = require('lodash/set');
var snakeCase = require('lodash/snakeCase');
var startCase = require('lodash/startCase');
var map = require('lodash/map');
var mapValues = require('lodash/mapValues');
var omitBy = require('lodash/omitBy');
var unset = require('lodash/unset');

/*----------------------------------------------
 *                    Kernel
 *---------------------------------------------- */
import env from './env';
import isArray from './isArray';
import isObject from './isObject';
import isEmpty from './isEmpty';
import isNotEmpty from './isNotEmpty';
import toBoolean from './toBoolean';
import removeEmpty from './removeEmpty';
import mapEmptyToNull from './mapEmptyToNull';

export { camelCase, capitalize, find, floor, get, has, hasIn, head, keysIn, last, set, snakeCase, startCase, map, mapValues, omitBy, unset, env, isArray, isObject, isEmpty, isNotEmpty, toBoolean, removeEmpty, mapEmptyToNull };
