// @kernel-js/exceptions/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.URIError = exports.TypeError = exports.TimeoutError = exports.SyntaxError = exports.SocketError = exports.ReferenceError = exports.RangeError = exports.NotSupportedError = exports.NotImplementedError = exports.NotFoundError = exports.InvalidOperationError = exports.HttpError = exports.ForbiddenError = exports.DataError = exports.ConnectionError = exports.AuthenticationRequiredError = exports.ArgumentNullError = exports.ArgumentError = exports.AlreadyInUseError = undefined;

var _ExtendableError = require('./ExtendableError');

var _ExtendableError2 = _interopRequireDefault(_ExtendableError);

var _AlreadyInUseError = require('./AlreadyInUseError');

var _AlreadyInUseError2 = _interopRequireDefault(_AlreadyInUseError);

var _ArgumentError = require('./ArgumentError');

var _ArgumentError2 = _interopRequireDefault(_ArgumentError);

var _ArgumentNullError = require('./ArgumentNullError');

var _ArgumentNullError2 = _interopRequireDefault(_ArgumentNullError);

var _AuthenticationRequiredError = require('./AuthenticationRequiredError');

var _AuthenticationRequiredError2 = _interopRequireDefault(_AuthenticationRequiredError);

var _ConnectionError = require('./ConnectionError');

var _ConnectionError2 = _interopRequireDefault(_ConnectionError);

var _DataError = require('./DataError');

var _DataError2 = _interopRequireDefault(_DataError);

var _ForbiddenError = require('./ForbiddenError');

var _ForbiddenError2 = _interopRequireDefault(_ForbiddenError);

var _HttpError = require('./HttpError');

var _HttpError2 = _interopRequireDefault(_HttpError);

var _InvalidOperationError = require('./InvalidOperationError');

var _InvalidOperationError2 = _interopRequireDefault(_InvalidOperationError);

var _NotFoundError = require('./NotFoundError');

var _NotFoundError2 = _interopRequireDefault(_NotFoundError);

var _NotImplementedError = require('./NotImplementedError');

var _NotImplementedError2 = _interopRequireDefault(_NotImplementedError);

var _NotSupportedError = require('./NotSupportedError');

var _NotSupportedError2 = _interopRequireDefault(_NotSupportedError);

var _RangeError = require('./RangeError');

var _RangeError2 = _interopRequireDefault(_RangeError);

var _ReferenceError = require('./ReferenceError');

var _ReferenceError2 = _interopRequireDefault(_ReferenceError);

var _SocketError = require('./SocketError');

var _SocketError2 = _interopRequireDefault(_SocketError);

var _SyntaxError = require('./SyntaxError');

var _SyntaxError2 = _interopRequireDefault(_SyntaxError);

var _TimeoutError = require('./TimeoutError');

var _TimeoutError2 = _interopRequireDefault(_TimeoutError);

var _TypeError = require('./TypeError');

var _TypeError2 = _interopRequireDefault(_TypeError);

var _URIError = require('./URIError');

var _URIError2 = _interopRequireDefault(_URIError);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.AlreadyInUseError = _AlreadyInUseError2.default;
exports.ArgumentError = _ArgumentError2.default;
exports.ArgumentNullError = _ArgumentNullError2.default;
exports.AuthenticationRequiredError = _AuthenticationRequiredError2.default;
exports.ConnectionError = _ConnectionError2.default;
exports.DataError = _DataError2.default;
exports.ForbiddenError = _ForbiddenError2.default;
exports.HttpError = _HttpError2.default;
exports.InvalidOperationError = _InvalidOperationError2.default;
exports.NotFoundError = _NotFoundError2.default;
exports.NotImplementedError = _NotImplementedError2.default;
exports.NotSupportedError = _NotSupportedError2.default;
exports.RangeError = _RangeError2.default;
exports.ReferenceError = _ReferenceError2.default;
exports.SocketError = _SocketError2.default;
exports.SyntaxError = _SyntaxError2.default;
exports.TimeoutError = _TimeoutError2.default;
exports.TypeError = _TypeError2.default;
exports.URIError = _URIError2.default;
exports.default = _ExtendableError2.default;
